diff --git a/calendar/base/content/calendar-alarm-dialog.js b/calendar/base/content/calendar-alarm-dialog.js
--- a/calendar/base/content/calendar-alarm-dialog.js
+++ b/calendar/base/content/calendar-alarm-dialog.js
@@ -97,18 +97,17 @@ function onDismissAllAlarms() {
  * Open the event dialog in the most recent Sunbird or Thunderbird window
  *
  * @param event     The itemdetails event.
  */
 function onItemDetails(event) {
     // We want this to happen in a calendar window.
     var wm = Components.classes["@mozilla.org/appshell/window-mediator;1"]
                        .getService(Components.interfaces.nsIWindowMediator);
-    var calWindow = wm.getMostRecentWindow("calendarMainWindow") ||
-                    wm.getMostRecentWindow("mail:3pane");
+    var calWindow = wm.getMostRecentWindow("collab:main");
     calWindow.modifyEventWithDialog(event.target.item, null, true);
 }
 
 /**
  * Sets up the alarm dialog, initializing the default snooze length and setting
  * up the relative date update timer.
  */
 var gRelativeDateUpdateTimer;
